lib/Connection.js

```diff
@@ -31,55 +33,80 @@ class Connection {
-						portOffset = 5 + addressLength;
-
-						requestBuffer.writeUInt16BE(this.port, portOffset, true);
-						this.socket.write(requestBuffer);
-						this.socket.once('data', (data) => {
-							if (data[1] !== 0) {
-								this.socket.end();
-								callback(`Socks5 connection request failed. Request denied, status: ${data[1]}`);
-							} else {
-								this.socket.setEncoding('utf8');
-								callback();
-							}
-						});
-					} else {
-						this.socket.end();
-						callback("Socks5 handshake failed. Closing the socket");
-					}
+
+		const doConnect = (onConnected) => {
+			if (this.proxyHost && this.proxyPort) {
+				const socks5Handshake = Buffer.alloc(3);
+				socks5Handshake[0] = 0x05; // SOCKS version number (must be 0x05 for this version)
+				socks5Handshake[1] = 0x01; // Number of authentication methods supported.
+				socks5Handshake[2] = 0x00; // 0x00: No authentication
+
+				this.socket.connect(this.proxyPort, this.proxyHost, () => {
+					// Socket connect done. Initiating Socks5 handshake
+					this.socket.write(socks5Handshake);
+
+					this.socket.once('data', (data) => {
+						if (data.toString('hex') === "0500") {
+							const addressLength = Buffer.byteLength(this.host);
+							const requestBuffer = Buffer.alloc(7 + addressLength);
+							let portOffset;
+
+							requestBuffer[0] = 0x05; // SOCKS version number (must be 0x05 for this version)
+							requestBuffer[1] = 0x01; // establish a TCP/IP stream connection
+							requestBuffer[2] = 0x00; // reserved, must be 0x00
+							requestBuffer[3] = 0x03; // address type, 1 byte. 0x03 = Domain name
+							requestBuffer[4] = addressLength; // 1 byte of name length followed by the name for domain name
+							requestBuffer.write(this.host, 5, addressLength);
+							portOffset = 5 + addressLength;
+
+							requestBuffer.writeUInt16BE(this.port, portOffset, true);
+							this.socket.write(requestBuffer);
+							this.socket.once('data', (data) => {
+								if (data[1] !== 0) {
+									this.socket.end();
+									callback(`Socks5 connection request failed. Request denied, status: ${data[1]}`);
+								} else {
+									this.socket.setEncoding('utf8');
+									onConnected();
+								}
+							});
+						} else {
+							this.socket.end();
+							onConnected("Socks5 handshake failed. Closing the socket");
+						}
+					});
 				});
-			});
-		} else {
-			this.socket.connect(this.port, this.host, callback);
-			this.socket.setEncoding('utf8');
-			this.socket.setTimeout(10000);
-		}
+			} else {
+				this.socket.connect(this.port, this.host, onConnected);
+				this.socket.setEncoding('utf8');
+				this.socket.setTimeout(10000);
+			}
+		};
+
+		doConnect((err) => {
+			if(err) {
+				callback(err);
+			}
+			if(this.sslOptions.enabled) {
+				this.createTlsConnection(callback, this.socket);
+			} else {
+				callback();
+			}
+		});
 		return this;
 	}
 
+	createTlsConnection(callback, preWrappedSocket = null) {
```

Comment:
```suggestion
	createTlsConnection(callback, socket = null) {
```

Can we not use "prewrappedSocket" the variable naming is clunky. "socket" is sufficient :)

Reply:
Thanks, now removed preWrappedSocket argument.